### ISSUE TITLE
Reduce verbosity during dotted key access.

### ISSUE THREAD

**Issue by ben-crowhurst** [marzer/tomlplusplus#118]
When accessing a property at a known dotted key location.  The current API requires splitting the key and iterating over the subscript operator in an attempt to get the value.

`return _config["cpu"]["watermark"]["upper"].value_or(default_value);`

It would be convenient and less verbose to write:

`return _config.find("cpu.watermark.upper");`

**Comment by marzer** [comment]
Hi, thanks for the suggestion. This indeed would be a good convenience utility to add to the library. I don't really have the time to do non-bugfix work on toml++, but fortunately in your case an implementation of this functionality would be pretty easy to stitch together locally in the mean time. The helper function I put together in my response to [this issue](https://github.com/marzer/tomlplusplus/issues/78) gets you most of the way there, only needing the string-splitting functionality and it's pretty much exactly what you're after.

**Comment by marzer** [comment]
@ben-crowhurst This is now implemented on the [v3 branch](https://github.com/marzer/tomlplusplus/tree/v3), as:
- free function `toml::at_path(node&, std::string_view)`
- member function `toml::node::at_path(std::string_view)`
- member function `toml::node_view::at_path(std::string_view)`

I didn't call it `find()` since that already has known semantics with `toml::table()` being analogous to `std::map`. Don't know if I love the name `at_path()` either, but naming things is hard ¯\\\_(ツ)\_/¯

The full list of changes in the v3 branch can be found [here](https://github.com/marzer/tomlplusplus/blob/v3/CHANGELOG.md).

**Comment by bencrowhurst** [comment]
Fantastic @marzer!  I'd hoped to get to this myself, but finding time is always a constraining factor.

Thank you for this feature and the project :+1:

**Comment by wakely** [comment]
The `at_path()` methods are really a nice addition.  

Is there a way to go the other way around?  In other words, from a `node_view`, extract the path that got us to it?  Essentially something equivalent to:

```c++
auto nv = table.at_path("here.is.my.path");
auto path = nv.path_to_node();  //returns "here.is.my.path"
```

**Comment by marzer** [comment]
@wakely Currently, no, there's no way to go back in the other direction. Doing so would require that nodes + node views keep track of their parents, which isn't the case in the current implementation.

I might consider adding this at some stage, but since it would necessitate adding additional state to nodes and node views, it will be an ABI break, so it won't be until the next time I do a major release.

If you're always using `at_path()` to get the node views in the first place you could implement a poor man's version of this yourself by wrapping `at_path()` in a function that returns a node view + path pair, e.g.:

```cpp

template <typename Node>
struct node_view_path_pair
{
	std::string path; // or std::string_view if the at_path() inputs are always string literals
	toml::node_view<Node> node_view;
}

node_view_path_pair<toml::node> at_path_retraceable(node& root, std::string_view path)
{
	auto result = toml::at_path(root, path);
	if (result)
		return { path, result };
	return {};
}
```

**Comment by wakely** [comment]
@marzer thanks for the feedback and suggestion.  The use case is really more for pathfinding after having arrived at a node by some other arcane table traversal voodoo.  I could wrap everything info methods that "accumulate" path, I guess, but this would have been crisp and clean.

**Comment by marzer** [comment]
> I could wrap everything info methods that "accumulate" path, I guess, but this would have been crisp and clean.

Yup, certainly would be cleaner as member functions. Until then there's a helper function example in the discussion of [this issue](https://github.com/marzer/tomlplusplus/issues/78) that might help with the accumulation implementation. It's not exactly what you want but would be easy to modify to suit.
